dart/test/protocol/rate_limiter_test.dart

```diff
@@ -0,0 +1,177 @@
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNotNull);
+    expect(1, result!.items.length);
+    //expect(2, result!.items.length); // TODO Update after added second item
+  });
+
+  test(
+      'When X-Sentry-Rate-Limit categories are empty, applies to all the categories',
+      () {
+    final rateLimiter = fixture.getSUT();
+    fixture.dateTimeToReturn = 0;
+    final eventItem = SentryEnvelopeItem.fromEvent(SentryEvent());
+    final envelope =
+        SentryEnvelope(SentryEnvelopeHeader.newEventId(), [eventItem]);
+
+    rateLimiter.updateRetryAfterLimits('50::key', null, 1);
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNull);
+  });
+
+  test(
+      'When all categories is set but expired, applies only for specific category',
+      () {
+    final rateLimiter = fixture.getSUT();
+    fixture.dateTimeToReturn = 0;
+    final eventItem = SentryEnvelopeItem.fromEvent(SentryEvent());
+    final envelope =
+        SentryEnvelope(SentryEnvelopeHeader.newEventId(), [eventItem]);
+
+    rateLimiter.updateRetryAfterLimits(
+        '1::key, 60:default;error;security:organization', null, 1);
+
+    fixture.dateTimeToReturn = 1001;
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNull);
+  });
+
+  test('When category has shorter rate limiting, do not apply new timestamp',
+      () {
+    final rateLimiter = fixture.getSUT();
+    fixture.dateTimeToReturn = 0;
+    final eventItem = SentryEnvelopeItem.fromEvent(SentryEvent());
+    final envelope =
+        SentryEnvelope(SentryEnvelopeHeader.newEventId(), [eventItem]);
+
+    rateLimiter.updateRetryAfterLimits(
+        '60:error:key, 1:error:organization', null, 1);
+
+    fixture.dateTimeToReturn = 1001;
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNull);
+  });
+
+  test('When category has longer rate limiting, apply new timestamp', () {
+    final rateLimiter = fixture.getSUT();
+    fixture.dateTimeToReturn = 0;
+    final eventItem = SentryEnvelopeItem.fromEvent(SentryEvent());
+    final envelope =
+        SentryEnvelope(SentryEnvelopeHeader.newEventId(), [eventItem]);
+
+    rateLimiter.updateRetryAfterLimits(
+        '1:error:key, 5:error:organization', null, 1);
+
+    fixture.dateTimeToReturn = 1001;
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNull);
+  });
+
+  test('When both retry headers are not present, default delay is set', () {
+    final rateLimiter = fixture.getSUT();
+    fixture.dateTimeToReturn = 0;
+    final eventItem = SentryEnvelopeItem.fromEvent(SentryEvent());
+    final envelope =
+        SentryEnvelope(SentryEnvelopeHeader.newEventId(), [eventItem]);
+
+    rateLimiter.updateRetryAfterLimits(null, null, 429);
+
+    fixture.dateTimeToReturn = 1001;
+
+    final result = rateLimiter.filter(envelope);
+    expect(result, isNull);
+  });
+}
+
+class Fixture {
+  var dateTimeToReturn = 0;
+
+  RateLimiter getSUT() {
```

Comment:
`dateTimeToReturn` could be an optional param with default value of `_currentDateTime`
so you don't need to mutate using `fixture.dateTimeToReturn = 1001;` but rather passing it to the `getSUT` call, unless you need to assert `dateTimeToReturn` which is not the case

Reply:
Doing it this way, as `RateLimiter` takes a closure for current date time. If i'd pass it as param, i'd also return a new instance of `RateLimiter`, which is not what i want in the tests. I want to be able to change the current date time but keep the instance around during the test.